### PR TITLE
Introduce debugWithActiveLayoutCleared to avoid duplicated code

### PR DESCRIPTION
**Pull request by fzyzcjy**
This is indeed a part of https://github.com/flutter/flutter/pull/113817. However, that one actually does two things in one PR: refactor internal code, and exposes an API. Thus, in order to follow the spirit mentioned in Flutter - one PR for only one thing - I make this refactor a separate PR, which should be easier to review.

### Performance difference

As is discussed in https://github.com/flutter/flutter/pull/113817#issue-1417868505 via compiler explorer, the generate assembly is not worse as long as we use that prefer-inline annotation.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -1582,6 +1582,26 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   static RenderObject? get debugActiveLayout => _debugActiveLayout;
   static RenderObject? _debugActiveLayout;
 
+  /// Set [debugActiveLayout] to null when [inner] callback is called.
+  /// This is useful when you have to temporarily clear that variable to
+  /// disable some false-positive checks, such as when computing toStringDeep
+  /// or using custom trees
+  @pragma('vm:prefer-inline')
+  static T _debugWithActiveLayoutCleared<T>(T Function() inner) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/object.dart:1590]: maybe name this `_withDebugActiveLayoutCleared`? Methods starting with `debug` or `_debug` should only be called from within an `assert` or guarded by kIsDebug. This doesn't have that restriction, though.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/object.dart (hunk 1 above)
Comment (fzyzcjy) [reply packages/flutter/lib/src/rendering/object.dart:1590]: Agree, renamed (will push code in a minute)

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -3418,18 +3427,8 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
-      return true;
-    }());
-    final String result = super.toStringShallow(joiner: joiner, minLevel: minLevel);
-    assert(() {
-      _debugActiveLayout = debugPreviousActiveLayout;
-      return true;
-    }());
-    return result;
+    return _debugWithActiveLayoutCleared(
+        () => super.toStringShallow(joiner: joiner, minLevel: minLevel));
```
[the remark this answers is not in the document]
Comment (fzyzcjy) [reply packages/flutter/lib/src/rendering/object.dart:3431]: done

**Review by justinmc — APPROVED** [review]
LGTM 👍 

There was a test failure the Windows framework tests.  I couldn't see why by glancing through the output, so I just reran it.  Let's see if it passes.

**Comment by fzyzcjy** [comment]
Hmm it still fails. I have seen this test failing across my PRs recently without reasons. let me bump ci

**Pull request merged**